File: Controllers/HomeController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Portfolio.Data;
using Portfolio.Models;
using Portfolio.Models.ViewModels;

namespace Portfolio.Controllers
{
    public class HomeController : Controller
    {
        private readonly PortfolioDbContext _context;
        private readonly ILogger<HomeController> _logger;

        public HomeController(PortfolioDbContext context, ILogger<HomeController> logger)
        {
            _context = context;
            _logger = logger;
        }

        // GET: / (Main Portfolio Website)
        [HttpGet]
        public async Task<IActionResult> Index([FromQuery] string? sent)
        {
            try
            {
                var admin = await _context.Admins.FirstOrDefaultAsync() ?? new Admin();
                var settings = await _context.Settings.FirstOrDefaultAsync() ?? new Settings();

                var skills = await _context.Skills
    .OrderBy(s => s.CategoryId)
    .ThenBy(s => s.Name)
    .ToListAsync();

                var projects = await _context.Projects
                    .Where(p => p.Status == "Published")
                    .OrderByDescending(p => p.IsFeatured)
                    .ThenByDescending(p => p.CreatedAt)
                    .ToListAsync();

                var courses = await _context.Courses
    .OrderByDescending(c => c.CreatedAt)
    .ToListAsync();

                var viewModel = new HomeViewModel
                {
                    Admin = admin,
                    Settings = settings,
                    Skills = skills,
                    Projects = projects,
                    Courses = courses,
                    TotalProjects = projects.Count,
                    TotalSkills = skills.Count,
                    TotalCourses = courses.Count,
                    MessageSuccess = (sent == "true")
                        ? "Thank you! Your message has been received successfully. I will get back to you shortly."
                        : null
                };

                ViewData["Title"] = $"{admin.Name} | {admin.Title}";
                return View(viewModel);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error fetching portfolio data from SQL Server.");
                return View(new HomeViewModel());
            }
        }

        // POST: /Home/SendMessage (or /send-message)
        [HttpPost]
        [ValidateAntiForgeryToken]
        public async Task<IActionResult> SendMessage([FromForm] Message model)
        {
            try
            {
                if (string.IsNullOrWhiteSpace(model.Name) || string.IsNullOrWhiteSpace(model.Email) || string.IsNullOrWhiteSpace(model.MessageText))
                {
                    if (Request.Headers["X-Requested-With"] == "XMLHttpRequest")
                    {
                        return Json(new { success = false, message = "Please fill in all required fields (Name, Email, Message)." });
                    }
                    return RedirectToAction(nameof(Index), new { fragment = "Contact" });
                }

                model.SentAt = DateTime.UtcNow;
                model.IsRead = false;
                if (string.IsNullOrWhiteSpace(model.Subject))
                {
                    model.Subject = "New Website Contact Inquiry";
                }

                _context.Messages.Add(model);
                await _context.SaveChangesAsync();

                if (Request.Headers["X-Requested-With"] == "XMLHttpRequest")
                {
                    return Json(new
                    {
                        success = true,
                        message = "Message submitted successfully! Thank you for reaching out.",
                        messageId = model.MessageID
                    });
                }

                return RedirectToAction(nameof(Index), new { sent = "true" });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error saving contact message to SQL Server.");
                if (Request.Headers["X-Requested-With"] == "XMLHttpRequest")
                {
                    return StatusCode(500, new { success = false, message = "Database error saving message. Please try again." });
                }
                return RedirectToAction(nameof(Index));
            }
        }

        // POST: /Home/EnrollCourse (or /api/enroll)
        [HttpPost]
        public async Task<IActionResult> EnrollCourse([FromForm] CourseEnrollment model, [FromForm] IFormFile? proofFile)
        {
            try
            {
                if (model.CourseID <= 0 || string.IsNullOrWhiteSpace(model.StudentName) || string.IsNullOrWhiteSpace(model.StudentEmail))
                {
                    return Json(new { success = false, message = "Please select a course and provide your name and email." });
                }

                var course = await _context.Courses.FindAsync(model.CourseID);
                if (course == null)
                {
                    return Json(new { success = false, message = "Selected course not found." });
                }

                model.CourseTitle = course.Title;
                model.EnrolledAt = DateTime.UtcNow;

                // Handle file upload if provided
                if (proofFile != null && proofFile.Length > 0)
                {
                    var uploadsFolder = Path.Combine(Directory.GetCurrentDirectory(), "wwwroot", "Images");
                    if (!Directory.Exists(uploadsFolder)) Directory.CreateDirectory(uploadsFolder);

                    var uniqueFileName = Guid.NewGuid().ToString() + "_" + Path.GetFileName(proofFile.FileName);
                    var filePath = Path.Combine(uploadsFolder, uniqueFileName);
                    using (var fileStream = new FileStream(filePath, FileMode.Create))
                    {
                        await proofFile.CopyToAsync(fileStream);
                    }
                    model.ScreenshotProof = uniqueFileName;
                }

                // If course is free, approve immediately
                if (course.PriceType?.ToLower() == "free" || course.Price?.ToLower() == "free")
                {
                    model.Status = "Approved";
                    model.AccessCode = "FREE-" + Guid.NewGuid().ToString().Substring(0, 8).ToUpper();
                }
                else
                {
                    model.Status = "Pending";
                    model.AccessCode = "TRX-" + (model.TransactionId ?? Guid.NewGuid().ToString().Substring(0, 8).ToUpper());
                }

                _context.CourseEnrollments.Add(model);

                // Increment enrolled count
                course.EnrolledStudents += 1;
                _context.Courses.Update(course);

                await _context.SaveChangesAsync();

                return Json(new
                {
                    success = true,
                    message = model.Status == "Approved"
                        ? "Free access granted! You can begin learning immediately."
                        : "Enrollment application submitted to Daniyal Khan! Your access will be approved after payment confirmation.",
                    enrollment = model
                });
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error processing course enrollment.");
                return StatusCode(500, new { success = false, message = "Internal error processing enrollment." });
            }
        }

        // GET: /Home/CheckEnrollment
        [HttpGet]
        public async Task<IActionResult> CheckEnrollment([FromQuery] string? query)
        {
            if (string.IsNullOrWhiteSpace(query))
            {
                return Json(new { success = false, message = "Please enter your Email or Transaction ID." });
            }

            var q = query.Trim().ToLower();
            var enrollments = await _context.CourseEnrollments
                .Include(e => e.Course)
                .Where(e => e.StudentEmail.ToLower() == q || (e.TransactionId != null && e.TransactionId.ToLower() == q))
                .OrderByDescending(e => e.EnrolledAt)
                .ToListAsync();

            return Json(new { success = true, enrollments });
        }

        // GET: /Home/GetCourseDetails
        [HttpGet]
        public async Task<IActionResult> GetCourseDetails([FromQuery] int id)
        {
            var course = await _context.Courses.FindAsync(id);
            if (course == null) return NotFound(new { success = false, message = "Course not found" });
            return Json(new { success = true, course });
        }
    }
}

File: src/controllers/homeController.js
import { db } from '../db/database.js';
import { askDaniyalAI } from '../utils/aiMentor.js';
import { sendEmailNotification } from '../utils/mailer.js';

export const homeController = {
  // Render Main Portfolio Website
  getIndex: (req, res) => {
    try {
      // Only show Published skills, projects, and active courses to the public
      const skills = db.getSkills().filter(s => (s.status || 'Published').toLowerCase() === 'published');
      const projects = db.getProjects().filter(p => (p.status || 'Published').toLowerCase() === 'published');
      
      // Free courses must appear at the top, followed by Paid courses
      const courses = db.getCourses()
        .filter(c => (c.status || 'Published').toLowerCase() !== 'draft')
        .sort((a, b) => {
          const aFree = a.price === 'Free' || a.priceType === 'free' || String(a.price).toLowerCase() === 'free';
          const bFree = b.price === 'Free' || b.priceType === 'free' || String(b.price).toLowerCase() === 'free';
          if (aFree && !bFree) return -1;
          if (!aFree && bFree) return 1;
          return (a.courseID || 0) - (b.courseID || 0);
        });

      const categories = db.getCategories();
      const admin = db.getAdmin();
      const settings = db.getSettings();
      const stats = db.getGlobalStats();

      const messageSuccess = req.query.sent === 'true' 
        ? 'Thank you! Your message has been received successfully. I will get back to you shortly.' 
        : null;

      res.render('home', {
        title: `${admin.name} | ${admin.title}`,
        skills,
        projects,
        courses,
        categories,
        admin,
        settings,
        stats,
        messageSuccess
      }, (err, html) => {
        if (err) {
          console.error('Render Home Error:', err);
          return res.status(500).send('Server Error Rendering Home Page');
        }
        res.render('layout', {
          title: `${admin.name} | ${admin.title}`,
          admin,
          settings,
          body: html
        });
      });
    } catch (error) {
      console.error('Home Page Controller Error:', error);
      res.status(500).send('Internal Server Error');
    }
  },

  // Submit Contact Message
  postSendMessage: async (req, res) => {
    try {
      const { name, email, phone, subject, messageText, message } = req.body;
      const actualMsg = messageText || message;
      if (!name || !email || !actualMsg) {
        if (req.xhr || req.headers['x-requested-with'] === 'XMLHttpRequest') {
          return res.status(400).json({ success: false, message: 'Please fill in all required fields.' });
        }
        return res.redirect('/#Contact');
      }

      const saved = db.saveMessage({
        name,
        email,
        phone: phone || '',
        subject: subject || 'New Website Contact Inquiry',
        messageText: actualMsg
      });

      // Dispatch notification email to inoxentdani09@gmail.com
      sendEmailNotification({
        type: 'inquiry',
        name,
        email,
        phone: phone || '',
        subject: subject || 'New Website Contact Inquiry',
        message: actualMsg
      }).catch(err => console.warn('Email notify error:', err));

      if (req.xhr || req.headers['x-requested-with'] === 'XMLHttpRequest') {
        return res.json({ 
          success: true, 
          message: 'Message submitted successfully! Thank you for reaching out.',
          messageId: saved.messageID 
        });
      }

      res.redirect('/?sent=true#Contact');
    } catch (error) {
      console.error('SendMessage Error:', error);
      if (req.xhr || req.headers['x-requested-with'] === 'XMLHttpRequest') {
        return res.status(500).json({ success: false, message: 'Error processing message.' });
      }
      res.redirect('/#Contact');
    }
  },

  // Public JSON APIs (compatible with original routes)
  getSkillsJson: (req, res) => {
    const category = req.query.category || null;
    const skills = db.getSkills(category).filter(s => (s.status || 'Published').toLowerCase() === 'published');
    res.json(skills);
  },

  getProjectsJson: (req, res) => {
    const category = req.query.category || null;
    const projects = db.getProjects(category).filter(p => (p.status || 'Published').toLowerCase() === 'published');
    res.json(projects);
  },

  getCoursesJson: (req, res) => {
    const filter = req.query.type || null;
    const courses = db.getCourses(filter).filter(c => (c.status || 'Published').toLowerCase() !== 'draft');
    res.json(courses);
  },

  getCourseDetailsJson: (req, res) => {
    const id = req.query.id;
    const course = db.getCourseById(id);
    if (!course) return res.status(404).json({ success: false, message: 'Course not found' });
    res.json({ success: true, course });
  },

  postEnrollCourse: async (req, res) => {
    try {
      const { courseId, courseID, studentName, studentEmail, studentPhone, paymentMethod, transactionId, trxId, screenshotProof } = req.body;
      const targetCourseId = courseId || courseID;
      
      if (!targetCourseId || !studentName || !studentEmail) {
        return res.status(400).json({ success: false, message: 'Please provide course, your name and email.' });
      }

      const enrollment = db.enrollInCourse({
        courseId: targetCourseId,
        studentName,
        studentEmail,
        studentPhone,
        paymentMethod: paymentMethod || 'JazzCash',
        transactionId: transactionId || trxId,
        screenshotProof: screenshotProof || ''
      });

      // Notify admin email of new enrollment
      sendEmailNotification({
        type: 'enrollment',
        name: studentName,
        email: studentEmail,
        phone: studentPhone,
        courseName: enrollment.courseName,
        amount: enrollment.amountPaid,
        paymentMethod: enrollment.paymentMethod,
        trxId: enrollment.transactionId
      }).catch(err => console.warn('Enrollment email notify error:', err));

      return res.json({
        success: true,
        message: enrollment.status === 'Approved' 
          ? 'Free access granted! You can begin learning immediately.' 
          : 'Enrollment application submitted to Daniyal Khan! Your access will be approved after payment confirmation.',
        enrollment
      });
    } catch (err) {
      console.error('Enroll Error:', err);
      return res.status(500).json({ success: false, message: 'Enrollment processing error. Please try again.' });
    }
  },

  getCheckEnrollment: (req, res) => {
    try {
      const query = req.query.query || req.query.email || req.query.trx;
      if (!query) {
        return res.status(400).json({ success: false, message: 'Please enter your email or Transaction ID.' });
      }
      const enrollments = db.getStudentEnrollments(query);
      return res.json({
        success: true,
        enrollments
      });
    } catch (err) {
      console.error('Check Enrollment Error:', err);
      return res.status(500).json({ success: false, message: 'Failed to look up enrollment.' });
    }
  },

  postAiChat: async (req, res) => {
    try {
      const { message, history } = req.body;
      if (!message || !message.trim()) {
        return res.status(400).json({ success: false, reply: 'Please provide a question or message.' });
      }
      const reply = await askDaniyalAI(message.trim(), history || []);
      return res.json({ success: true, reply });
    } catch (error) {
      console.error('AI Chat Error:', error);
      return res.status(500).json({ success: false, reply: 'AI service is temporarily busy. Please try again or ask on WhatsApp.' });
    }
  }
};
